File: cvdl-rs/src/local_storage.rs
extern crate directories;

use std::{
    io::Write,
    path::{Path, PathBuf},
};

use directories::ProjectDirs;

use crate::{
    data_schema::DataSchema, layout_schema::LayoutSchema, resume_data::ResumeData,
    resume_layout::ResumeLayout,
};

/// This module provides the abstractions for interacting with persistent storage.
/// The library follows the directory structure below:
///
/// projectdir/com.cvdl.cvdl/
/// ├── data/
///        ├── resumes
///             ├── resume1.json
///             ├── resume2.json
///        ├── data-schemas.json
///        ├── layout-schemas.json
///        |── resume-layouts.json
///
/// The resume.json files contain the resume information, as well as references to the
/// schema names.
///
/// This module provides 3 types of functionalities for all 4 data types:
///     1. List
///     2. Load
///     3. Save

// Initiation Function

pub struct LocalStorage {
    dir: String,
}

impl LocalStorage {
    pub fn new() -> LocalStorage {
        let project_dirs = ProjectDirs::from("com", "cvdl", "cvdl").unwrap();
        let dir = project_dirs
            .data_dir()
            .as_os_str()
            .to_str()
            .unwrap()
            .to_string();
        LocalStorage { dir }
    }

    pub fn custom_dir(dir: &str) -> LocalStorage {
        LocalStorage {
            dir: dir.to_string(),
        }
    }
}

impl LocalStorage {
    pub fn initiate_local_storage(&self) {
        let data_dir = Path::new(self.dir.as_str());
        // Create data_dir/resumes if it does not exist
        std::fs::create_dir_all(data_dir.join("resumes")).unwrap();
        // Create data_dir/data-schemas.json if it does not exist
        if let Ok(mut data_schemas_file) =
            std::fs::File::create_new(data_dir.join("data-schemas.json"))
        {
            data_schemas_file.write_all("[]".as_bytes()).unwrap();
        }
        // Create data_dir/layout-schemas.json if it does not exist
        if let Ok(mut layout_schemas_file) =
            std::fs::File::create_new(data_dir.join("layout-schemas.json"))
        {
            layout_schemas_file.write_all("[]".as_bytes()).unwrap();
        }
        // Create data_dir/resume-layouts.json if it does not exist
        if let Ok(mut resume_layouts_file) =
            std::fs::File::create_new(data_dir.join("resume-layouts.json"))
        {
            resume_layouts_file.write_all("[]".as_bytes()).unwrap();
        }
    }
}

// Listing Functions

impl LocalStorage {
    pub fn list_resumes(&self) -> Vec<String> {
        let data_dir = Path::new(self.dir.as_str()).join("resumes");
        let mut resumes = Vec::new();
        if let Ok(entries) = std::fs::read_dir(data_dir) {
            for entry in entries.flatten() {
                if let Some(extension) = entry.path().extension() {
                    if extension == "json" {
                        if let Some(file_name) = entry.path().file_stem() {
                            resumes.push(file_name.to_str().unwrap().to_string());
                        }
                    }
                }
            }
        }
        resumes
    }

    pub fn list_data_schemas(&self) -> Vec<String> {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("data-schemas.json")).unwrap();
        let data_schemas: Vec<DataSchema> = serde_json::from_reader(file).unwrap();
        data_schemas
            .iter()
            .map(|schema| schema.schema_name.to_string())
            .collect()
    }

    pub fn list_layout_schemas(&self) -> Vec<String> {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("layout-schemas.json")).unwrap();
        let data_schemas: Vec<LayoutSchema> = serde_json::from_reader(file).unwrap();
        data_schemas
            .iter()
            .map(|schema| schema.schema_name.to_string())
            .collect()
    }

    pub fn list_resume_layouts(&self) -> Vec<String> {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("resume-layouts.json")).unwrap();
        let data_schemas: Vec<ResumeLayout> = serde_json::from_reader(file).unwrap();
        data_schemas
            .iter()
            .map(|schema| schema.schema_name.to_string())
            .collect()
    }
}

// Loading Functions

impl LocalStorage {
    pub fn load_resume(&self, resume_name: &str) -> ResumeData {
        let data_dir = Path::new(self.dir.as_str()).join("resumes");
        let file = std::fs::File::open(data_dir.join(format!("{}.json", resume_name))).unwrap();
        let resume_data: ResumeData = serde_json::from_reader(file).unwrap();
        resume_data
    }

    pub fn load_data_schema(&self, schema_name: &str) -> DataSchema {
        let data_dir = Path::new(self.dir.as_str());
        println!("{:?}", data_dir);
        let file = std::fs::File::open(data_dir.join("data-schemas.json")).unwrap();
        let data_schemas: Vec<DataSchema> = serde_json::from_reader(file).unwrap();
        data_schemas
            .into_iter()
            .find(|schema| schema.schema_name == schema_name)
            .unwrap()
    }

    pub fn load_layout_schema(&self, schema_name: &str) -> LayoutSchema {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("layout-schemas.json")).unwrap();
        let layout_schemas: Vec<LayoutSchema> = serde_json::from_reader(file).unwrap();
        layout_schemas
            .into_iter()
            .find(|schema| schema.schema_name == schema_name)
            .unwrap()
    }

    pub fn load_resume_layout(&self, schema_name: &str) -> ResumeLayout {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("resume-layouts.json")).unwrap();
        let resume_layout: Vec<ResumeLayout> = serde_json::from_reader(file).unwrap();
        resume_layout
            .into_iter()
            .find(|schema| schema.schema_name == schema_name)
            .unwrap()
    }
}

// Saving Functions

impl LocalStorage {
    pub fn save_resume(&self, resume_name: &str, resume_data: &ResumeData) {
        let data_dir = Path::new(self.dir.as_str()).join("resumes");
        let file = std::fs::File::create(data_dir.join(format!("{}.json", resume_name))).unwrap();
        serde_json::to_writer_pretty(file, resume_data).unwrap();
    }

    pub fn save_data_schema(&self, data_schema: &DataSchema) {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("data-schemas.json")).unwrap();
        let mut data_schemas: Vec<DataSchema> = serde_json::from_reader(file).unwrap();
        if let Some(index) = data_schemas
            .iter()
            .position(|schema| schema.schema_name == data_schema.schema_name)
        {
            data_schemas[index] = data_schema.clone();
        } else {
            data_schemas.push(data_schema.clone());
        }
        let file = std::fs::File::create(data_dir.join("data-schemas.json")).unwrap();
        serde_json::to_writer_pretty(file, &data_schemas).unwrap();
    }

    pub fn save_layout_schema(&self, layout_schema: &LayoutSchema) {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("layout-schemas.json")).unwrap();
        let mut layout_schemas: Vec<LayoutSchema> = serde_json::from_reader(file).unwrap();
        if let Some(index) = layout_schemas
            .iter()
            .position(|schema| schema.schema_name == layout_schema.schema_name)
        {
            layout_schemas[index] = layout_schema.clone();
        } else {
            layout_schemas.push(layout_schema.clone());
        }
        let file = std::fs::File::create(data_dir.join("layout-schemas.json")).unwrap();
        serde_json::to_writer_pretty(file, &layout_schemas).unwrap();
    }

    pub fn save_resume_layout(&self, resume_layout: &ResumeLayout) {
        let data_dir = Path::new(self.dir.as_str());
        let file = std::fs::File::open(data_dir.join("resume-layouts.json")).unwrap();
        let mut resume_layouts: Vec<ResumeLayout> = serde_json::from_reader(file).unwrap();
        if let Some(index) = resume_layouts
            .iter()
            .position(|schema| schema.schema_name == resume_layout.schema_name)
        {
            resume_layouts[index] = resume_layout.clone();
        } else {
            resume_layouts.push(resume_layout.clone());
        }
        let file = std::fs::File::create(data_dir.join("resume-layouts.json")).unwrap();
        serde_json::to_writer_pretty(file, &resume_layouts).unwrap();
    }
}

File: cvdl-ts/src/LocalStorage.ts
import { DataSchema } from "./DataSchema";
import { LayoutSchema } from "./LayoutSchema";
import * as Resume from "./Resume";
import { ResumeLayout } from "./ResumeLayout";
import { Storage } from "./Storage";
import * as Defaults from "./Defaults";

export class LocalStorage implements Storage {
	prefix: string;
	constructor() {
		this.prefix = "https://d2bnplhbawocbk.cloudfront.net/data/";
	}

	save_bindings(bindings: Map<string, unknown>): Promise<void> {
		const bindingsObject: any = {};
		for (const [key, value] of bindings) {
			bindingsObject[key] = value;
		}
		localStorage.setItem("bindings", JSON.stringify(bindingsObject));
		return Promise.resolve();
	}

	async initiate_storage(): Promise<void> {
		if (!localStorage.getItem("resumes")) {
			localStorage.setItem("resumes", JSON.stringify([Defaults.DefaultResume]));
		}

		if (!localStorage.getItem("data_schemas")) {
			localStorage.setItem("data_schemas", JSON.stringify(Defaults.DefaultDataSchemas));
		}

		if (!localStorage.getItem("section_layouts")) {
			localStorage.setItem(
				"section_layouts",
				JSON.stringify(Defaults.DefaultLayoutSchemas),
			);
		}

		if (!localStorage.getItem("resume_layouts")) {
			localStorage.setItem(
				"resume_layouts",
				JSON.stringify([Defaults.DefaultResumeLayout]),
			);
		}
	}

	list_resumes(): Promise<string[]> {
		const resumes = JSON.parse(localStorage.getItem("resumes") || "[]").map(
			(resume: any) => resume.name,
		);
		return Promise.resolve(resumes);
	}

	list_data_schemas(): Promise<string[]> {
		const schemas = JSON.parse(
			localStorage.getItem("data_schemas") || "[]",
		).map((schema: any) => schema.schema_name);
		return Promise.resolve(schemas);
	}

	list_layout_schemas(): Promise<string[]> {
		const schemas = JSON.parse(
			localStorage.getItem("section_layouts") || "[]",
		).map((schema: any) => schema.schema_name);
		return Promise.resolve(schemas);
	}

	list_resume_layouts(): Promise<string[]> {
		const schemas = JSON.parse(
			localStorage.getItem("resume_layouts") || "[]",
		).map((schema: any) => schema.schema_name);
		return Promise.resolve(schemas);
	}

	load_resume(resume_name: string): Promise<Resume.t> {
		const resume = JSON.parse(localStorage.getItem("resumes") || "[]").find(
			(resume: any) => resume.name === resume_name,
		);
		if (!resume) {
			throw new Error(`Resume(${resume_name}) not found`);
		}
		return Promise.resolve(resume);
	}

	load_data_schema(schema_name: string): Promise<DataSchema.t> {
		const schema = JSON.parse(
			localStorage.getItem("data_schemas") || "[]",
		).find((schema: any) => schema.schema_name === schema_name);
		if (!schema) {
			throw new Error(`Data Schema(${schema_name}) not found`);
		}
		return Promise.resolve(schema);
	}

	load_bindings(): Promise<Map<string, unknown>> {
		if (!localStorage.getItem("bindings")) {
			localStorage.setItem("bindings", JSON.stringify({}));
		}
		const bindingsObject = JSON.parse(localStorage.getItem("bindings") || "{}");
		const bindings = new Map<string, unknown>();
		for (const [key, value] of Object.entries(bindingsObject)) {
			bindings.set(key, value);
		}
		return Promise.resolve(bindings);
	}

	load_layout_schema(schema_name: string): Promise<LayoutSchema> {
		const schema = JSON.parse(
			localStorage.getItem("section_layouts") || "[]",
		).find((schema: any) => schema.schema_name === schema_name);
		if (!schema) {
			throw new Error(`Layout Schema(${schema_name}) not found`);
		}
		return Promise.resolve(LayoutSchema.fromJson(schema));
	}
	load_resume_layout(schema_name: string): Promise<ResumeLayout> {
		const schema = JSON.parse(
			localStorage.getItem("resume_layouts") || "[]",
		).find((schema: any) => schema.schema_name === schema_name);
		if (!schema) {
			throw new Error(`Resume Layout(${schema_name}) not found`);
		}
		return Promise.resolve(ResumeLayout.fromJson(schema));
	}
	save_resume(resume_name: string, resume_data: Resume.t): Promise<void> {
		const resumes = JSON.parse(localStorage.getItem("resumes") || "[]");
		let resumeIndex = resumes.findIndex((resume: any) => resume.name === resume_name);
		if (resumeIndex === -1) {
			resumes.push(resume_data);
		} else {
			resumes[resumeIndex] = resume_data;
		}
		localStorage.setItem("resumes", JSON.stringify(resumes));

		return Promise.resolve();
	}
	save_data_schema(data_schema: DataSchema.t): Promise<void> {
		const schemas = JSON.parse(
			localStorage.getItem("data_schemas") || "[]",
		);
		const schema = schemas.find(
			(schema: DataSchema.t) => schema.schema_name === data_schema.schema_name,
		);
		if (!schema) {
			schemas.push(data_schema);
		} else {
			schema.header_schema = data_schema.header_schema;
			schema.item_schema = data_schema.item_schema;
		}
		localStorage.setItem(
			"data_schemas",
			JSON.stringify(schemas),
		);

		return Promise.resolve();
	}
	save_layout_schema(layout_schema: LayoutSchema): Promise<void> {
		const schemasDirectMapped = JSON.parse(
			localStorage.getItem("section_layouts") || "[]",
		);
		const schemas = schemasDirectMapped.map((schema: any) =>
			LayoutSchema.fromJson(schema),
		);
		const schema = schemas.find(
			(schema: LayoutSchema) =>
				schema.schema_name === layout_schema.schema_name,
		);
		if (!schema) {
			schemas.push(layout_schema);
		} else {
			schema.header_layout_schema = layout_schema.header_layout_schema;
			schema.item_layout_schema = layout_schema.item_layout_schema;
		}
		localStorage.setItem(
			"section_layouts",
			JSON.stringify(schemas.map((schema: LayoutSchema) => schema.toJson())),
		);

		return Promise.resolve();
	}
	save_resume_layout(resume_layout: ResumeLayout): Promise<void> {
		throw new Error("Method not implemented.");
	}
	async load_font(fontName: string): Promise<Buffer> {
		const path = `fonts/${fontName}.ttf`;
		if (!localStorage.getItem(path)) {
			const response = await fetch(
				`https://d2bnplhbawocbk.cloudfront.net/data/${path}`,
			);
			const font_data = await response.arrayBuffer();
			return Buffer.from(font_data);
		}

		throw new Error("Font not found");
	}
}
